test/aeso_compiler_tests.erl

```diff
@@ -834,6 +837,173 @@ failing_contracts() ->
+                     "Trying to implement or extend an undefined interface `Z`">>
+                  ])
+    , ?TYPE_ERROR(contract_interface_polymorphism_same_decl_multi_interface,
+                  [<<?Pos(7,10)
+                     "Unimplemented function `f` from the interface `I` in the contract `C`">>
+                  ])
+    , ?TYPE_ERROR(contract_interface_polymorphism_same_name_same_type,
+                  [<<?Pos(7,10)
+                     "Unimplemented function `f` from the interface `I1` in the contract `C`">>
+                  ])
+    , ?TYPE_ERROR(contract_interface_polymorphism_same_name_different_type,
+                  [<<?Pos(9,5)
+                     "Duplicate definitions of `f` at\n"
+                     "  - line 8, column 5\n"
+                     "  - line 9, column 5">>
+                  ])
+    , ?TYPE_ERROR(contract_polymorphism_missing_implementation,
+                  [<<?Pos(7,10)
+                     "Unimplemented function `f` from the interface `I1` in the contract `C`">>
+                  ])
+    , ?TYPE_ERROR(contract_polymorphism_same_decl_multi_interface,
+                  [<<?Pos(7,10)
+                     "Unimplemented function `f` from the interface `J` in the contract `C`">>
+                  ])
+    , ?TYPE_ERROR(contract_polymorphism_undefined_interface,
+                  [<<?Pos(1,14)
+                     "Trying to implement or extend an undefined interface `I`">>
+                  ])
+    , ?TYPE_ERROR(contract_polymorphism_same_name_different_type_multi_interface,
+                  [<<?Pos(9,5)
+                     "Duplicate definitions of `f` at\n"
+                     "  - line 8, column 5\n"
+                     "  - line 9, column 5">>
+                  ])
+    , ?TYPE_ERROR(contract_interface_polymorphism_undefined_interface,
+                  [<<?Pos(1,24)
+                     "Trying to implement or extend an undefined interface `H`">>
+                  ])
+    , ?TYPE_ERROR(polymorphism_variance_switching,
+                  [<<?Pos(38,49)
+                     "Cannot unify `Cat` and `Animal`\n"
+                     "when checking the application of\n"
+                     "  `g2 : (Cat) => Cat`\n"
+                     "to arguments\n"
+                     "  `x : Animal`">>,
+                   <<?Pos(41,43)
+                     "Cannot unify `Animal` and `Cat`\n"
+                     "when checking the type of the expression `g3(x) : Animal` against the expected type `Cat`">>,
+                   <<?Pos(50,55)
+                     "Cannot unify `Animal` and `Cat`\n"
+                     "when checking the application of\n"
+                     "  `g5 : ((Animal) => Animal) => Cat`\n"
+                     "to arguments\n"
+                     "  `x : (Cat) => Cat`">>
+                  ])
+    , ?TYPE_ERROR(polymorphism_variance_switching_custom_types,
+                  [<<?Pos(56,32)
+                     "Cannot unify `Cat` and `Animal`\n"
```

Comment:
The error messages here could be more elaborate. I can already see people scratching their heads why Cat does not unify with Animal, but sometimes the other way around. I think it would be valuable to mention the variance in the message:
```
Cannot unify A with B in a contravariant context
```
Or something. Even if one does not know the concept of variance, they know at least what to google.

Reply:
Would it be better to say that `Animal` is not a subtype of `Cat` instead of mentioning the variance part? Someone would scratch his head saying that why didn't the compiler check if `Cat` is a subtype of `Animal` instead of checking if `Animal` is a subtype of `Cat, but people will always scratch their heads for some reason, and I don't see that mentioning the type of variance making it easier for people to find out what went wrong.

Reply:
Done that here da6d22717cf3410b97792894aba371f353078ea1